x/leverage/types/position.go

```diff
@@ -0,0 +1,230 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var minimumBorrowFactor = sdk.MustNewDecFromStr("0.5")
+
+// AccountPosition must be created by NewAccountPosition for proper initialization.
+// Contains an account's borrowed and collateral values, arranged into special asset
+// pairs and regular assets. Each list will always be sorted by collateral weight.
+// Also caches some relevant values that will be reused in computation, like token
+// settings and the borrower's total borrowed value and collateral value.
+type AccountPosition struct {
+	// all special asset pairs which apply to the account. Specifically, this should
+	// cache any special pairs which match the account's collateral, even if they do
+	// not match one of its borrows. This is the widest set of information that could
+	// be needed when calculating maxWithdraw of an already collateralized asset, or
+	// maxBorrow of any asset (even if not initially present in the position).
+	// A pair that does not currently apply to the account (but could do so if a borrow
+	// were added or existing pairs were rearranged) will have zero USD value
+	// but will still be initialized with the proper weights, and sorted in order.
+	specialPairs WeightedSpecialPairs
+	// all collateral value not being used for special asset pairs
+	collateral WeightedDecCoins
+	// all borrowed value not being used for special asset pairs
+	borrowed WeightedDecCoins
+	// caches retrieved token settings
+	tokens map[string]Token
+	// collateral value (using PriceModeLow and interpreting unknown prices as zero)
+	collateralValue sdk.Dec
+	// borrowed value (using PriceModeHigh and requiring all prices known)
+	borrowedValue sdk.Dec
+}
+
+// NewAccountPosition creates and sorts an account position based on token settings,
+// special asset pairs, and the collateral and borrowed value of each token in an account.
+// Once this structure is created, borrow limit calculations can be performed without
+// keeper or context.
+func NewAccountPosition(
+	tokens []Token,
+	pairs []SpecialAssetPair,
+	collateralValue,
+	borrowedValue sdk.DecCoins,
+) AccountPosition {
+	position := AccountPosition{
+		specialPairs:    WeightedSpecialPairs{},
+		collateral:      WeightedDecCoins{},
+		borrowed:        WeightedDecCoins{},
+		tokens:          map[string]Token{},
+		collateralValue: sdk.ZeroDec(),
+		borrowedValue:   sdk.ZeroDec(),
+	}
+
+	// cache all registered tokens
```

Comment:
This doesn't go beyond the query or transaction that is using `AccountPosition`, so it's not the kind of cache that causes gas nondeterminism.

We would get only tokens represented in the existing position, but this struct must be capable of computing max borrow for assets that aren't initially present as well, so we fetch all.